search_gov_crawler/scrapy_scheduler.py

```diff
@@ -0,0 +1,106 @@
+from search_gov_crawler.search_gov_spiders.extensions.json_logging import LOG_FMT
+from search_gov_crawler.search_gov_spiders.utility_files.init_schedule import SpiderSchedule
+
+logging.basicConfig(level=logging.INFO)
+log = logging.getLogger()
+log.handlers[0].setFormatter(JsonFormatter(fmt=LOG_FMT))
+
+CRAWL_SITES_FILE = Path(__file__).parent / "search_gov_spiders" / "utility_files" / "crawl-sites.json"
+
+
+def run_scrapy_crawl(spider: str, allowed_domains: str, start_urls: str) -> None:
+    """Runs `scrapy crawl` command as a subprocess given the allowed arguments"""
+
+    scrapy_env = os.environ.copy()
+    scrapy_env["PYTHONPATH"] = str(Path(__file__).parent.parent)
+
+    subprocess.run(
+        f"scrapy crawl {spider} -a allowed_domains={allowed_domains} -a start_urls={start_urls}",
+        check=True,
+        cwd=Path(__file__).parent,
+        env=scrapy_env,
+        executable="/bin/bash",
+        shell=True,
+    )
+    log.info(
+        "Successfully completed scrapy crawl with args spider=%s, allowed_domains=%s, start_urls=%s",
+        spider,
+        allowed_domains,
+        start_urls,
+    )
+
+
+def transform_crawl_sites(crawl_sites: list[dict]) -> list[dict]:
+    """Transform crawl sites records into a format that can be used to"""
+
+    transformed_crawl_sites = []
+    schedule = SpiderSchedule()
+
+    for crawl_site in crawl_sites:
+        job_name = str(crawl_site["name"])
+        schedule_slot = schedule.get_next_slot()
+        transformed_crawl_sites.append(
+            {
+                "func": run_scrapy_crawl,
+                "id": job_name.lower().replace(" ", "-").replace("---", "-"),
+                "name": job_name,
+                "trigger": CronTrigger(
+                    year="*",
+                    month="*",
+                    day="*",
+                    week="*",
+                    day_of_week=schedule_slot.day,
+                    hour=schedule_slot.hour,
+                    minute=schedule_slot.minute,
+                    second=0,
+                    timezone="UTC",
+                    jitter=0,
+                ),
+                "args": [
+                    "domain_spider" if not crawl_site["handle_javascript"] else "domain_spider_js",
+                    crawl_site["allowed_domains"],
+                    crawl_site["starting_urls"],
+                ],
+            }
+        )
+
+    return transformed_crawl_sites
+
+
+def start_scrapy_scheduler(input_file: Path) -> None:
+    """Initializes schedule from input file, schedules jobs and runs scheduler"""
+
+    # Load and transform crawl sites
+    crawl_sites = json.loads(input_file.read_text(encoding="UTF-8"))
+    apscheduler_jobs = transform_crawl_sites(crawl_sites)
+
+    # Initalize scheduler
+    scheduler = BlockingScheduler(
+        jobstores={"memory": MemoryJobStore()},
+        executors={"default": ThreadPoolExecutor(max_workers=5)},
+        job_defaults={"coalesce": False, "max_instances": 1},
```

Comment:
Should we maybe max these out to the instance's available resources. Not sure how it works, but usually if you don't provide the params (max_workers and max_instances) it will use the system's max

Reply:
For max workers... looks like it defaults (for apscheduler at least) to 10 https://apscheduler.readthedocs.io/en/stable/modules/executors/pool.html?highlight=threadpoolexecutor#apscheduler.executors.pool.ThreadPoolExecutor. Sure we could increase should it be a environment variable we could set upstream in ansible or something?  
```python
executors={"default": ThreadPoolExecutor(max_workers=os.environ.get("AP_SCHEDULER_MAX_WORKERS", 10))}
```
default could be 10 or 20 or whatever but i prefer an explicit call out here even if we go with 10.

for max_instances, i couldn't find it documented but in the BaseScheduler class it appears the default is 1:
https://github.com/agronholm/apscheduler/blob/2a2210fc71683547f8b01c7d4552815c3b67c1fe/apscheduler/schedulers/base.py#L714

This is how many instances of the crawl job can run at once.  This seems like it should be left at 1, right?  Do we want to be running a crawl of example.com and then start another example.com crawl before that first one finishes?

The other day I did some manual testing of what happens when you hit these numbers and apscheduler just logs a warning with a reason and skips the run.

Reply:
Ah, okay makes sense. Yeah, I think 10 is a good start then. Usually ec2 instances have 2 threads per core so we can assume: `max_workers=instance_cpus*2`

Reply:
i made the update to hardcode to 10... if we want to adjust or parameterize later we can do that.